File: bridges/voiceover/Sources/VoiceOverBridgeAdapters/FileTranscript.swift
// ROLE: adapter implementing the Transcript port; it owns the transcript's vocabulary, one timestamped line per event.
// USED BY: the Session, through the port. BUILT BY: Wiring, via `session(in:)`.
// Line shapes match the NVDA bridge's transcript, so a tester reads one format for both readers.

import Foundation
import VoiceOverBridgeDomain

public final class FileTranscript: Transcript {
	private let writer: any FileWriter
	private let timestamp: () -> String
	private var isOpen = false

	public var logPath: String { writer.path }

	public init(writer: any FileWriter, timestamp: @escaping () -> String = FileTranscript.wallclock) {
		self.writer = writer
		self.timestamp = timestamp
	}

	public func open() {
		try? writer.open()
		isOpen = true
	}

	public func sessionOpened(mode: String, voice: String, persona: String) {
		// `-` when absent, so every SESSION OPEN line carries the same fields.
		line("SESSION OPEN mode=\(mode) voice=\(voice) persona=\(persona.isEmpty ? "-" : persona)")
	}

	/// Quoted: the words come from another process, and an unescaped newline would forge a transcript line.
	public func speech(_ text: String) {
		line("SPEECH \(FileTranscript.quoted(text))")
	}

	public func gesture(_ command: String) {
		line("GESTURE \(FileTranscript.quoted(command))")
	}

	/// Only the length is recorded, never the typed text.
	public func typed(_ length: Int) {
		line("TYPE length=\(length)")
	}

	public func announced(_ text: String) {
		line("ANNOUNCE \(FileTranscript.quoted(text))")
	}

	public func note(_ text: String) {
		line("NOTE \(text)")
	}

	public func sessionClosed(reason: String) {
		line("SESSION CLOSE reason=\(reason)")
		isOpen = false
		writer.close()
	}

	private func line(_ text: String) {
		guard isOpen else { return }
		writer.writeLine("\(timestamp()) \(text)")
	}

	static func quoted(_ text: String) -> String {
		let escaped =
			text
			.replacingOccurrences(of: "\\", with: "\\\\")
			.replacingOccurrences(of: "\"", with: "\\\"")
			.replacingOccurrences(of: "\n", with: "\\n")
			.replacingOccurrences(of: "\r", with: "\\r")
		return "\"\(escaped)\""
	}

	public static func wallclock() -> String {
		let formatter = DateFormatter()
		formatter.locale = Locale(identifier: "en_US_POSIX")
		formatter.dateFormat = "yyyy-MM-dd HH:mm:ss.SSS"
		return formatter.string(from: Date())
	}
}

public extension FileTranscript {
	static func defaultLogDirectory(home: String) -> String {
		URL(fileURLWithPath: home).appendingPathComponent("Library/Logs/screen-readers-mcp").path
	}

	/// Keeps the newest `keep` transcripts; the stamp in each name sorts lexically in time order.
	static func session(in directory: String, keep: Int = 20, stamp: String? = nil) -> FileTranscript {
		let manager = FileManager.default
		try? manager.createDirectory(atPath: directory, withIntermediateDirectories: true)
		let name = stamp ?? {
			let formatter = DateFormatter()
			formatter.locale = Locale(identifier: "en_US_POSIX")
			formatter.dateFormat = "yyyyMMdd-HHmmss-SSS"
			return formatter.string(from: Date())
		}()
		let existing = ((try? manager.contentsOfDirectory(atPath: directory)) ?? [])
			.filter { $0.hasPrefix("session-") && $0.hasSuffix(".log") }
			.sorted()
		for stale in existing.dropLast(max(0, keep - 1)) {
			try? manager.removeItem(atPath: URL(fileURLWithPath: directory).appendingPathComponent(stale).path)
		}
		return FileTranscript(
			writer: TextFileWriter(
				path: URL(fileURLWithPath: directory).appendingPathComponent("session-\(name).log").path
			)
		)
	}
}

File: bridges/nvda/addon/globalPlugins/nvdaMcpBridge/adapters/file_transcript.py
# nvdaMcpBridge adapters -- FileTranscript: the Transcript vocabulary.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
# ROLE: adapter implementing the Transcript port over the FileWriter seam.
# USED BY: the Session controller.
# BUILT BY: create_session_log.

from __future__ import annotations

import os
from collections.abc import Callable
from datetime import datetime
from pathlib import Path
from typing import TYPE_CHECKING

from ..domain.ports.transcript import Transcript
from .text_file_writer import TextFileWriter

if TYPE_CHECKING:
	from .ports.file_writer import FileWriter

DEFAULT_KEEP: int = 20


def _wallclock() -> str:
	return datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]


class FileTranscript(Transcript):
	def __init__(self, writer: FileWriter, *, timestamp: Callable[[], str] = _wallclock) -> None:
		self._writer = writer
		self._timestamp = timestamp
		self._open = False

	@property
	def path(self) -> str:
		return self._writer.path

	def open(self) -> None:
		self._writer.open()
		self._open = True

	def _line(self, text: str) -> None:
		if not self._open:
			return
		self._writer.write_line(f"{self._timestamp()} {text}")

	def session_opened(self, mode: str, synth: str, persona: str) -> None:
		# An absent persona is written as `-` so every SESSION OPEN line has the same shape.
		self._line(f"SESSION OPEN mode={mode} synth={synth} persona={persona or '-'}")

	def gesture(self, gesture_id: str) -> None:
		self._line(f"GESTURE {gesture_id}")

	def typed(self, length: int) -> None:
		self._line(f"TYPE length={length}")

	def speech(self, text: str) -> None:
		self._line(f"SPEECH {text!r}")

	def note(self, text: str) -> None:
		self._line(f"NOTE {text}")

	def session_closed(self, reason: str) -> None:
		self._line(f"SESSION CLOSE reason={reason}")
		self._open = False
		self._writer.close()


def create_session_log(
	logs_dir: str | os.PathLike[str],
	*,
	keep: int = DEFAULT_KEEP,
	timestamp: Callable[[], str] = _wallclock,
	name_stamp: Callable[[], str] | None = None,
) -> FileTranscript:
	directory = Path(logs_dir)
	directory.mkdir(parents=True, exist_ok=True)
	stamp = (name_stamp or (lambda: datetime.now().strftime("%Y%m%d-%H%M%S-%f")))()
	transcript = FileTranscript(TextFileWriter(directory / f"session-{stamp}.log"), timestamp=timestamp)
	transcript.open()
	_prune(directory, keep)
	return transcript


def _prune(directory: Path, keep: int) -> None:
	# Names embed a time-sortable stamp, so a lexical sort is chronological.
	existing = sorted(directory.glob("session-*.log"), key=lambda p: p.name)
	for stale in existing[: max(0, len(existing) - keep)]:
		try:
			stale.unlink()
		except OSError:
			pass
